plugins/mongodb/mongodbembed.cpp

```diff
@@ -0,0 +1,2287 @@
+/** \example mongodb-test.ecl
+ * Here is some example code of the plugin being used.
+ */
+namespace mongodbembed
+{
+    const time_t OBJECT_EXPIRE_TIMEOUT_SECONDS = 60 * 2; // Two minutes
+    static std::once_flag CONNECTION_CACHE_INIT_FLAG;
+
+    //--------------------------------------------------------------------------
+    // Plugin Classes
+    //--------------------------------------------------------------------------
+
+    /**
+     * @brief Takes an exception object and outputs the error to the user
+     * 
+     * @param e mongocxx::exception has some values that are useful to look for
+     */
+    void reportQueryFailure(const mongocxx::exception &e)
+    {
+        if (e.code() == mongocxx::error_code::k_invalid_collection_object) 
+        {
+            failx("Collection not found: %s",e.what());
+        }
+        if (e.code().value() == 26) 
+        {
+            failx("NamespaceNotFound: %s",e.what());
+        }
+        if (e.code().value() == 11000) 
+        {
+            failx("Duplicate Key: %s",e.what());
+        }
+        if (!isEmptyString(e.what()))
+        {
+            failx("Exception: %s", e.what());
+        }
+    }
+
+    /**
+     * @brief deserialize MongoDB Extended JSON relaxed format into normal JSON for the RowStream class
+     * Examples of EJSON relaxed format are:
+     * "dateField": {"$date":{"$numberLong":"1565546054692"}}
+     * "doubleField": {"$numberDouble":"10.5"}
+     * For more documentation on EJSON. https://www.mongodb.com/docs/manual/reference/mongodb-extended-json/
+     * 
+     * @param result Reference to a result string where the standard JSON should be written.
+     * @param row Pointer to the beginning of the result row.
+     */
+    void deserializeEJSON(std::string &result, const char * row)
+    {
+        auto start = row;
+        int depth = 0; // Keeps track of depth so we don't exit on wrong '}'
+        const char * lastBrkt;
+
+        while (*start && (*start != '}' || depth > 1)) 
+        {
+            if (*start == '\"')
+            {
+                // Find end of string ignoring any quotes inside the string. This is done by checking what comes after the string which is not foolproof if someone
+                // matches json standard output inside their string.
+                while(*start && !(*start == '\"' && (*(start + 1) == ' ' || *(start + 1) == ',') && (*(start + 2) == ':' || *(start + 2) == '}' || *(start + 2) == ']')))
+                {
+                    start++;
+                }
+                start++;
+            }
+            if (*start == '{') // If we see a document increase the depth
+            {
+                depth++;
+                lastBrkt = start; // Keep track of last bracket in case we need to backtrack
+            }
+            if (*start == '{' && *(start + 1) == ' ' && *(start + 2) == '\"' && *(start + 3) == '$') // Look for "{ \"$" to mark the start of a datatype
```

Comment:
it feels like this block should be within the the if block starting in L161 (for readability/maintainability)

Reply:
I agree. I will refactor that code to make it more readable because currently it is dense and not clear what the purpose behind it is.